src/netbase.cpp

```diff
@@ -68,17 +79,49 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
     while (ai_trav != nullptr) {
         if (ai_trav->ai_family == AF_INET) {
             assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
+            req->vAddr->emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
         }
         if (ai_trav->ai_family == AF_INET6) {
             assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in6));
             const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
-            resolved_addresses.emplace_back(s6->sin6_addr, s6->sin6_scope_id);
+            req->vAddr->emplace_back(s6->sin6_addr, s6->sin6_scope_id);
         }
         ai_trav = ai_trav->ai_next;
     }
     freeaddrinfo(ai_res);
 
+    LOCK(gai_mutex);
+    req->complete = true;
+}
+
+std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_lookup)
+{
+    std::vector<CNetAddr> resolved_addresses;
+    struct GAIRequest req = {.allow_lookup = allow_lookup, .name = name, .vAddr = &resolved_addresses};
+
+    if (allow_lookup) {
+        // Execute the DNS lookup in another thread and check it periodically for completion.
+        // After sufficiently waiting, abandon the thread entirely. The user's system
+        // may wait forever for a DNS response that never returns.
+        std::thread thread(AsyncGetAddrInfo, &req);
+        int checks{0};
+        while (true)
+        {
+            {
+                LOCK(gai_mutex);
+                if (req.complete) break;
+            }
+            // Check every 100ms for 2 seconds
+            if (++checks > 20) break;
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+        thread.detach();
```

Comment:
Afaict this won't work here, because the thread might not finish and then the pool would just get clogged (see comment about abandoning the thread).

Performance should also not really matter here since this doesn't happen that often and startup time for a thread vs. time of a DNS lookup is probably negligible.

Reply:
> Afaict this won't work here, because the thread might not finish and then the pool would just get clogged (see comment about abandoning the thread).

k, `getaddrinfo` doesn't have a timeout and `getaddrinfo_a` seems to have a segfault. Cool..

> Performance should also not really matter here since this doesn't happen that often and startup time for a thread vs. time of a DNS lookup is probably negligible.

I was thinking more about context switching. Isn't this used for `OpenNetworkConnection` too? So, it would create a new thread for every connection attempt? (unless the lookups are numeric, which guess that is the rule here..).

And, following the worst case scenario, what if the user call `addnode` a good number of times? we would get a lot of detached threads? (unless we set a limit for the amount of lookups that can be resolved in parallel).